_posts/2020-06-23-Durr_Hoyer_Post_1.md

```diff
@@ -0,0 +1,323 @@
+
+```
+   // intialize register to number of qubits as there are indices
+   using ((Register) = (Qubit[TableLength])) 
+   { 
+      // Create Uniform Superposition of all indices
+      PrepareUniformSuperposition(TableLength,LittleEndian(Register)); 
+      // Grab the qubit in the RandomIndex (which is intialized in our python host script using numpy.uniform.random()) and set it aside              
+      let Marker = Register[RandomIndex]; 
+      // Apply Oracle to flip all states that are T[j]<T[y]
+      Controlled Z(Register,Marker); 
+   }
+```
+
+Step 2(a) has been stasified. Now we must figure out how to apply the QESA algorithm. 
+For the following circuits assume our Random Index is 0. 
+#### QESA
+The QESA algorithm is a generalized Grover's search characterized by the following circuit for an even number of table elements.     
+
+![2 items in a list](/assets/images/2-item-list.gif)
+If we have an odd number of table elements we use this circuit, where QFT is the approixmate Fourier transform as given by Kitaev [(3)].    
+
+![3 items in a list](/assets/images/3-item-list.gif)
+
+Lets break down whats going on here.
+We utilize the register we were working with earlier and apply a H transform, this is stated to simply be the Hadmard. 
+
+```
+   // Apply Hadamard to register
+   ApplyToEach(H,Register); 
+```
+
+We then apply a conditional phase shift if the qubit is 0.    
+```
+   // Reflect qubits that are 0s
+   ApplyConditionalPhase_0(LittleEndian(Register));  
+   
+   operation ApplyConditionalPhase_0(register: LittleEndian) : Unit is Adj + Ctl
+   {
+       using (aux = Qubit()) 
+       {
+           (ControlledOnInt(0,Z))(register!,aux); // If qubit is 0 flip it!
+       }
+   }
+```
+From here we apply the inverse of the Hadmard, this is just the conjugate transpose since Hadamard is unitary. 
+This can be implemented be calling the Adjunct of Hadamard.  
+
+```
+   // Apply Adjunct Hadamard to register
+   ApplyToEachA(H,Register); 
+```
+The last step is another conditional phase shift, though it is applied if the qubit is 1.
+
+```
+   // Reflect qubits that are 1s
+   ApplyConditionalPhase(LittleEndian(Register)); 
+   
+   operation ApplyConditionalPhase(register : LittleEndian) : Unit is Adj + Ctl 
+   {
+       using (aux = Qubit()) 
+       {
+           (ControlledOnInt(1,Z))(register!,aux); // If qubit is 1 flip it!
+       }
+   }
+```
+#### QESA Script
+Our Q# script will be structured as follows:
+
+```
+
+    namespace QESA {
+        open Microsoft.Quantum.Intrinsic;
+        open Microsoft.Quantum.Diagnostics;
+        open Microsoft.Quantum.Arrays; 
+        open Microsoft.Quantum.Preparation;
+        open Microsoft.Quantum.Canon;
+        open Microsoft.Quantum.Arithmetic;
+        open Microsoft.Quantum.Math;
+        open Microsoft.Quantum.Core;
+        open Microsoft.Quantum.Convert;
+        operation Algorithm_Even(TableLength : Int, RandomIndex : Int) : Unit

```

Comment:
Not expecting Unit here, because the algorithm expects to get an index.
Probably the result we be a result (or result array).